src/webui/webapplication.cpp

```diff
@@ -173,45 +108,149 @@ namespace
+            else {
+                found = false; // no more translatable strings
+            }
+        }
+    }
+
+#ifdef Q_OS_UNIX
+    bool isRegularFile(const char *path)
+    {
+        struct ::stat st;
+        if (!::stat(path, &st)) {
+            //  analyse erno and log the error
+            const auto err = errno;
+            qDebug("Could not get file stats for path '%s'. Error: %s"
+                   , qUtf8Printable(path), qUtf8Printable(strerror(err)));
+            return false;
+        }
+
+        return (st.st_mode & S_IFMT) == S_IFREG;
+    }
+#endif
 }
 
-void WebApplication::action_public_index()
+WebApplication::WebApplication(QObject *parent)
+    : AbstractWebApplication(parent)
+    , m_isAltUIUsed(false)
 {
-    QString path;
-
-    if (!args_.isEmpty()) {
-        if (args_.back() == "favicon.ico")
-            path = ":/icons/skin/qbittorrent16.png";
-        else
-            path = WWW_FOLDER + args_.join("/");
+    const QLatin1String tagPublic {"WEBAPI_PUBLIC"};
+    const QLatin1String tagPrivate {"WEBAPI_PRIVATE"};
+    for (int i = metaObject()->methodOffset(); i < metaObject()->methodCount(); ++i) {
+        QMetaMethod method = metaObject()->method(i);
+        if (method.tag() == tagPublic)
+            m_publicAPIs << method.name();
+        else if (method.tag() == tagPrivate)
+            m_privateAPIs << method.name();
     }
 
-    printFile(path);
+    configure();
+    connect(Preferences::instance(), &Preferences::changed, this, &WebApplication::configure);
 }
 
-void WebApplication::action_public_webui()
+void WebApplication::sendWebUIFile()
 {
-    if (!sessionActive())
-        printFile(PRIVATE_FOLDER + "login.html");
-    else
-        printFile(PRIVATE_FOLDER + "index.html");
+    const QStringList pathItems {request().path.split('/', QString::SkipEmptyParts)};
+    if (pathItems.contains(".") || pathItems.contains("..")) {
+        status(500, "Internal Server Error");
+        return;
+    }
+
+    if ((!m_isAltUIUsed) && (request().path.startsWith(PATH_PREFIX_IMAGES))) {
+        sendFile(QLatin1String(":/icons/") + request().path.mid(PATH_PREFIX_IMAGES.size()));
+        header(Http::HEADER_CACHE_CONTROL, MAX_AGE_MONTH);
+        return;
+    }
+
+    const QString path {
+        (request().path != QLatin1String("/")
+                ? request().path
+                : (sessionActive()
+                   ? QLatin1String("/index.html")
+                   : QLatin1String("/login.html")))
+    };
+
+    QString localPath {
+        m_rootFolder
+                + (sessionActive() ? PRIVATE_FOLDER : PUBLIC_FOLDER)
+                + path
+    };
+
+    QFileInfo fileInfo {localPath};
+
+    if (!fileInfo.exists() && sessionActive()) {
+        // try to send public file if there is no private one
+        localPath = m_rootFolder + PUBLIC_FOLDER + path;
+        fileInfo.setFile(localPath);
+    }
+
+#ifdef Q_OS_UNIX
+    if (!isRegularFile(localPath.toUtf8().constData())) {
```

Comment:
Another problem, built-in webUI won't load because it won't pass this check. 
I got the "Unacceptable file type." msg in my browser.

Reply:
Maybe the function is used with Qt resources?

Reply:
In fact, all these checks are unnecessary for the builtin UI.
Will be fixed later.
@evsh, or we shouldn't trust ourself as well?

Reply:
> @evsh, or we shouldn't trust ourself as well?

:D

Reply:
@Chocobo1, try it now!

Reply:
Thanks, now I can pass this check.